OpenRA.Mods.Common/Graphics/ContrailRenderable.cs

```diff
@@ -66,11 +67,24 @@ public void Render(WorldRenderer wr)
 			// Start of the first line segment is the tail of the list - don't smooth it.
 			var curPos = trail[Index(next - skip - 1)];
 			var curColor = color;
-			for (var i = 0; i < length - skip - 4; i++)
+
+			for (var i = 1; i < renderLength; i++)
 			{
-				var j = next - skip - i - 2;
-				var nextPos = Average(trail[Index(j)], trail[Index(j - 1)], trail[Index(j - 2)], trail[Index(j - 3)]);
-				var nextColor = Exts.ColorLerp(i * 1f / (length - 4), color, Color.Transparent);
+				var j = next - skip - 1 - i;
+				WPos nextPos;
+
+				// Smooth the contrail to tail direction, use 4 contrail position for max smooth effect.
+				if (i < renderLength - 3)
+					nextPos = Average(trail[Index(j)], trail[Index(j - 1)], trail[Index(j - 2)], trail[Index(j - 3)]);
+				else
+				{
+					var smoothparams = new List<WPos>();
```

Comment:
Creating a new list every render is a really bad idea wrt garbage generation. Its not immediately clear what this is trying to do, but it should be possible to rewrite this as a LINQ statement using `Reverse`, `Skip`, `Take` and `Average`.

Reply:
`trail` is a loop list, and it needs accessing by `Index()`. Simply `Take()` cannot meet the requirement of this loop list

Reply:
not very good at Linq, any other idea?

Reply:
Use LINQ to build the indices, then do `.Select(i => trail[i]).Average()`.

Reply:
Errr... how to build the indices in a simple way then?

Like now, we need to choose number `n` in `k < renderLength - i && k < maxSmoothLength`, and translate the number `n` to `index => Index(n)`, then use `.Select(i => trail[i]).Average()`

I think if the code use LINQ is more complicated than those code below, we should think twice before we use LINQ.
`
				int nextX = 0, nextY = 0, nextZ = 0, k = 0;
				for (; k < renderLength - i && k < maxSmoothLength; k++)
				{
					nextX += trail[Index(j - k)].X;
					nextY += trail[Index(j - k)].Y;
					nextZ += trail[Index(j - k)].Z;
				}

				nextPos = new WPos(nextX / k, nextY / k, nextZ / k);
`

Reply:
That code won't work as-is: it will suffer from integer overflow. Take a look at how `WPos.Average` uses `long` for the intermediate values, and add a comment explaining that this is manually inlined from `WPos.Average` to simplify averaging within the circular buffer.

Reply:
Please also try to make an effort to match our standard coding style for things like `var` instead of explicit types, and comments being on their own lines. You've worked with OpenRA code enough now that these should be clear!

Reply:
Thanks, but is there any idea on LINQ of it? It is better to reuse `Average`.

Reply:
I think you are right about skipping LINQ and manually inlining the averaging code. This just needs to be done properly and documented.

Reply:
Fixed.
Coding one's own `Average` here still feels not good, since we have got an `Average` already. I will miss LINQ here.